### PR TITLE
Added provider "darkanime"

### PR DESCRIPTION
**Pull request by ArjixGamer**

### CODE REVIEW

**Comment by ArjixGamer** [comment]
This is an attempt of a provider, it works fine and all but it can take a bit more polish than what i gave it. Only the mp4upload source is supported because all the other sources are broken most of the time

**Comment by MichaelDavidJohnson** [comment]
Don't merge this until he learns to not put multiple imports on one line.

File: anime_downloader/sites/darkanime.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import json, re
```
Comment (MichaelDavidJohnson) [inline anime_downloader/sites/darkanime.py:1]: ...

File: anime_downloader/sites/darkanime.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import json, re
+
+from anime_downloader.sites.anime import Anime, AnimeEpisode, SearchResult
+from anime_downloader.sites import helpers
+
+class DarkAnime(Anime, sitename = 'darkanime'):
+
```
Comment (MichaelDavidJohnson) [inline anime_downloader/sites/darkanime.py:7]: Too much whitespace

File: anime_downloader/sites/darkanime.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            SearchResult(
+                title = i[0],
+                url = i[1],
+                )
+            for i in results
+            ]
+
+        return search_results
+
+
```
Comment (MichaelDavidJohnson) [inline anime_downloader/sites/darkanime.py:26]: Whitespace everywhere

File: anime_downloader/sites/darkanime.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    def _scrape_metadata(self):
+            self.title = helpers.soupify(helpers.get(self.url).text).find_all('h2')[0].text.strip()
+class DarkAnimeEpisode(AnimeEpisode, sitename='darkanime'):
+    def _get_sources(self):
+        soup = helpers.soupify(helpers.get(self.url).text)
+        players = soup.find_all('iframe')
+        sources = soup.find_all('script')[-3].string
+        regex = r"\[([^)]+)\]"
+        sources = json.loads('[' + re.search(regex, sources).group(1) + ']')
+        link = 'https://www.mp4upload.com/embed-{}.html'.format(sources[-1]['source']) if sources[-1]['host'] == 'mp4upload' else 'https://www.mp4upload.com/embed-{}.html'.format(sources[0]['source'])
```
Comment (MichaelDavidJohnson) [inline anime_downloader/sites/darkanime.py:43]: Kinda not a fan of the single line logic here, it's just too long and is annoying to read. That's probably just me though

File: anime_downloader/sites/darkanime.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+class DarkAnime(Anime, sitename = 'darkanime'):
+
+    sitename = 'darkanime'
+
+    @classmethod
+    def search(cls, query):
+
+        soup = helpers.soupify(helpers.get('https://app.darkanime.stream/api/v1/animes', params={'term': query}).json()['animesHtml'])
+        soup = soup.find_all('a', href=True)
+        results = [[x.find('h3').text.strip(), 'https://app.darkanime.stream' + x['href']] for x in soup]
```
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:15]: Why do you define results and loop through that instead of doing searchresult() directly on the info?

File: anime_downloader/sites/darkanime.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        eps.reverse()
+        return eps
+
+    def _scrape_metadata(self):
+            self.title = helpers.soupify(helpers.get(self.url).text).find_all('h2')[0].text.strip()
+class DarkAnimeEpisode(AnimeEpisode, sitename='darkanime'):
+    def _get_sources(self):
+        soup = helpers.soupify(helpers.get(self.url).text)
+        players = soup.find_all('iframe')
+        sources = soup.find_all('script')[-3].string
```
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:40]: I'd prefer regex or at least a loop to make this more stable. This would break if they just did one small unrelated script change on their site.

File: anime_downloader/sites/darkanime.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        regex = r"\[([^)]+)\]"
+        sources = json.loads('[' + re.search(regex, sources).group(1) + ']')
```
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:42]: Instead of adding [ and ] you can just move the parentheses to capture the brackets too.

File: anime_downloader/sites/darkanime.py (hunk 5 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:15]: because i like that

File: anime_downloader/sites/darkanime.py (hunk 6 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:40]: i will try

File: anime_downloader/sites/darkanime.py (hunk 7 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:42]: can you explain more or give the regex? i really am bad with it

File: anime_downloader/sites/darkanime.py (hunk 3 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:26]: its just whitespace, live with it

File: anime_downloader/sites/darkanime.py (hunk 2 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:7]: its just whitespace, live with it

File: anime_downloader/sites/darkanime.py (hunk 1 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:1]: reeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee

File: anime_downloader/sites/darkanime.py (hunk 4 above)
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:43]: Just taking a glance at the site I'm pretty sure you can almost copy-paste all of DreamAnimeEpisode from dreamanime and have it working. Even better would be if you added sort_sources() to it.

File: anime_downloader/sites/darkanime.py (hunk 7 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:42]: ok, i got it

File: anime_downloader/sites/darkanime.py (hunk 4 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:42]: yep, with some edits it works, i dont know if i will add sort_sources()

Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py]: fixed by following a suggestion by lagrad

File: anime_downloader/sites/darkanime.py (hunk 3 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py]: i already fixed it, dont react like that lagrad

**Comment by ArjixGamer** [comment]
i think its done

File: anime_downloader/sites/darkanime.py [hunk 8]
```diff
@@ -41,18 +41,23 @@ def getLink(self, name, _id):
 
     def _get_sources(self):
         server = self.config.get("server", "mp4upload")
-        resp = helpers.soupify(helpers.get(self.url).text).find_all('script')[-3].string
-        hosts = json.loads(re.search(r"(\[[^)]+\])", resp).group(1))
+        resp = helpers.soupify(helpers.get(self.url).text).find_all('script')#[-3].string
+        for a in resp:
+            try:
```
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:46]: Remove this try except at all costs. You said it's due to NoneType, then do something more like
```
for i in resp:
	if i:
		if 'sources' in i.string:
			res = i.string
```

Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:46]: have you tried it urself? even with ur suggestion i get ERROR: argument of type 'NoneType' is not iterable
and i already tried this

File: anime_downloader/sites/darkanime.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        html = helpers.soupify(helpers.get(self.url).text)
+        eps = html.find('ul', class_='mt-4').find_all('li')
+        eps = ['https://app.darkanime.stream' + x.a['href'] for x in eps]
+        eps.reverse()
+        return eps
+    
+    def _scrape_metadata(self):
+            self.title = helpers.soupify(helpers.get(self.url).text).find_all('h2')[0].text.strip()
+            
+class DarkAnimeEpisode(AnimeEpisode, sitename='darkanime'):
```
Comment (Blatzar) [inline anime_downloader/sites/darkanime.py:33]: ill do a sort sources later

File: anime_downloader/sites/darkanime.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        elif name == "mp4upload":
+            return f"https://mp4upload.com/embed-{_id}.html"
+        elif name == "xstreamcdn":
+            return "https://www.xstreamcdn.com/v/" + _id
+
+    def _get_sources(self):
+        server = self.config.get("server", "mp4upload")
+        resp = helpers.soupify(helpers.get(self.url).text).find_all('script')#[-3].string
+        for a in resp:
+            if type(a.string) != type(None):
```
[the remark this answers is not in the document]
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:46]: should i change it then?

File: anime_downloader/sites/darkanime.py (hunk 9 above)
Comment (ArjixGamer) [reply anime_downloader/sites/darkanime.py:33]: woah, big brain time, i did find_all()[0] bc it has only 1 result, but i could just do a find()
